---
editor_options:
  chunk_output_type: console
output:
  pdf_document: default
  html_document: default
---
 
# starting files: gene counts . (UMI have been treated)

---
title: "R script for the analyzing ghita's data with only Irf1_FF and Ikbkb_FF with resolution 1.3"
author: "Ammar Sabir Cheema"
date: "03/04/2020"
output: html_document
---
  
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

<br>
# loading libraries:
```{r load libraries}
library(Seurat)
library(dplyr)
library(devtools)
library(ggrepel)
library(uwot)
library(scales)
library(pheatmap)
```


# Reading and processing input files for only floxed condition
```{r}

metadata <- read.table("metadata5_wo_M1_wo_conta.txt", sep="\t", header=T, row.names=1)
dim(metadata)

metadata1 <- metadata[which(metadata$Genotype == "Ikbkb_FF" | metadata$Genotype == "Irf1_FF" ),]
write.table(metadata1, "metadata1_ghita_data_only_Irf1_FF_only_Ikbkb_FF_res_1.3.txt", sep="\t", quote=F)
dim(metadata1)

y1 <- read.table("y4_gene_expression_wo_M1_wo_conta.txt", sep="\t", header=T, row.names=1)
y2 <- as.data.frame(t(y1))
y3 = y2[rownames(y2) %in% paste('X',rownames(metadata1),sep=""), ,drop=F]
y4 <- as.data.frame(t(y3))
write.table(y4,"y4_only_Irf1_FF_only_Ikbkb_FF_res_1.3.txt",sep="\t", quote=F)
dim(y4)


```

# Reading the processed metadata file
```{r}
metadata2 <- read.table("metadata1_ghita_data_only_Irf1_FF_only_Ikbkb_FF_res_1.3.txt", sep="\t", header=T, row.names=1)

```

# Making the Seurat object after removing the cells which are not Irf1_FF and Ikbkb_FF

```{r}
y_seurat <- CreateSeuratObject(y4, min.cells = 5, min.features = 500, project = "Ghita_data_single_cell_removed_Irf1_FF_Ikbkb_FF")

dim(y_seurat)

```


# Performing Some Quality Checks (QC)
```{r}
VlnPlot(object = y_seurat, features = c("nCount_RNA", "nFeature_RNA"))


hist(y_seurat@meta.data$nCount_RNA)
hist(y_seurat@meta.data$nFeature_RNA)
```

# Subsetting the metadata object according to the newly generated seurat object
```{r}
metadata2 = metadata1[paste("X",rownames(metadata1),sep = "") %in%  colnames(y_seurat), , drop=F]
write.table(metadata2, "metadata2_only_Irf1_FF_only_Ikbkb_FF_res_1.3.txt", sep="\t", quote=F)
dim(metadata2)
```

# Performing some quality checks using the metadata information
```{r}
df_counts <- as.data.frame(y_seurat@meta.data$nCount_RNA)
df_counts$phenotype <- metadata2$Genotype
df_counts$plate <- metadata2$Plate
df_counts$experiment <- metadata2$experiment
df_counts$mouse_age <- metadata2$mouse_age
df_counts$id <- rownames(metadata2) # these are cells (id)


# histogram number of reads, according to plate
ggplot(df_counts, aes(x=id, y=y_seurat@meta.data$nCount_RNA, fill=as.factor(metadata2$Plate))) + geom_col(stat="identity") + theme(axis.text.x = element_blank())

# histogram number of reads, according to genotype
ggplot(df_counts, aes(x=id, y=y_seurat@meta.data$nCount_RNA, fill=as.factor(metadata2$Genotype))) + geom_col(stat="identity") + theme(axis.text.x = element_blank())

# histogram number of reads, according to experiment
ggplot(df_counts, aes(x=id, y=y_seurat@meta.data$nCount_RNA, fill=as.factor(metadata2$experiment))) + geom_col(stat="identity") + theme(axis.text.x = element_blank())

# histogram number of reads, according to mouse_age
ggplot(df_counts, aes(x=id, y=y_seurat@meta.data$nCount_RNA, fill=as.factor(metadata2$mouse_age))) + geom_col(stat="identity") + theme(axis.text.x = element_blank())

# histogram number of genes/features, according to plate
ggplot(df_counts, aes(x=id, y=y_seurat@meta.data$nFeature_RNA, fill=as.factor(metadata2$Plate))) + geom_col(stat="identity") + theme(axis.text.x = element_blank())

# histogram number of genes/features, according to phenotype
ggplot(df_counts, aes(x=id, y=y_seurat@meta.data$nFeature_RNA, fill=as.factor(metadata2$Genotype))) + geom_col(stat="identity") + theme(axis.text.x = element_blank())

# histogram number of genes/features, according to experiment
ggplot(df_counts, aes(x=id, y=y_seurat@meta.data$nFeature_RNA, fill=as.factor(metadata2$experiment))) + geom_col(stat="identity") + theme(axis.text.x = element_blank())

# histogram number of genes/features, according to mouse_age
ggplot(df_counts, aes(x=id, y=y_seurat@meta.data$nFeature_RNA, fill=as.factor(metadata2$mouse_age))) + geom_col(stat="identity") + theme(axis.text.x = element_blank())
```

## Normalization the data set
```{r Normalization in natural log y_seurat}
norm.data <- NormalizeData(object = y_seurat, normalization.method = "LogNormalize", scale.factor = 1e6)
write.table(norm.data@assays$RNA@data, "log_norm_filtered_data__only_Irf1_FF_only_Ikbkb_FF_res_1.3.txt", sep="\t", quote=F)
```

# Identification of highly variable features (feature selection)
# Here we identify 1000 most variable features and plot 40 of them on the plot shown below
# Detect features that are variable on a feature plot
```{r}
norm.data <- FindVariableFeatures(object=norm.data, selection.method = "vst", mean.function = ExpMean, dispersion.function = LogVMR, num.bin = 20, binning.method = "equal_width", dispersion.cutoff=c(1, Inf), mean.cutoff=c(3, 12), nfeatures = 1000, verbose = TRUE)


# Identify the 40 most highly variable genes
top40 <- head(VariableFeatures(norm.data), 40)

# plot variable features with and without labels
# We plot the variable features in the same way as we find the Variable features in the above method
plot1 <- VariableFeaturePlot(norm.data)

# In plot2 we label the plot1 with top40 variable genes we have detected above
plot2 <- LabelPoints(plot = plot1, points = top40, repel = TRUE)
#CombinePlots(plots = list(plot1, plot2))
plot2
```



# Scaling the dataset
```{r}
all.genes <- rownames(norm.data)
norm.data <- ScaleData(norm.data, features = all.genes)
```



## Perform linear dimensional reduction (PCA and ICA)
```{r}
norm.data <- RunPCA(norm.data, npcs = 30, features = VariableFeatures(object = norm.data))
norm.data <- RunICA(norm.data, nics = 30, features = VariableFeatures(object = norm.data))
```

# Examine and visualize PCA results a few different ways
```{r}
print(norm.data[["pca"]], dims = 1:5, nfeatures = 15)
print(norm.data[["ica"]], dims = 1:5, nfeatures = 15)

VizDimLoadings(norm.data, dims = 1:2, reduction = "pca")
VizDimLoadings(norm.data, dims = 1:2, reduction = "ica")

# Plot showing Principle componenet 1 (PC1) and Principle componenet 2 (PC2)
DimPlot(norm.data, dims=c(1, 2), reduction = "pca", pt.size=2)
# Plot showing Principle componenet 3 (PC3) and Principle componenet 4 (PC4)
DimPlot(norm.data, dims=c(3, 4), reduction = "pca", pt.size=2)
# Plot showing Principle componenet 5 (PC5) and Principle componenet 6 (PC6)
DimPlot(norm.data, dims=c(5, 6), reduction = "pca", pt.size=2)


pca_coordinates <- norm.data@reductions$pca@cell.embeddings
#write.table(pca_coordinates, "PCA_coordinates.txt", sep="\t", quote=F)
ica_coordinates <- norm.data@reductions$ica@cell.embeddings
#write.table(ica_coordinates, "ICA_coordinates.txt", sep="\t", quote=F)
```

# Heatmap:
```{r}
DimHeatmap(norm.data, dims = 1, cells = 113, nfeatures = 20, balanced = TRUE)
DimHeatmap(norm.data, dims = 2, cells = 113, balanced = TRUE)

DimHeatmap(norm.data, dims = 1:9, cells = 113, balanced = TRUE)
```

# Get cell and feature names, and total numbers
```{r}
head(colnames(x = norm.data))
head(Cells(norm.data))
head(rownames(x = norm.data))
ncol(x = norm.data)
nrow(x = norm.data)
```

# Determine the ‘dimensionality’ of the dataset
# From the plot shown below it is clear that first four principle components are responsible for most of the variability
```{r}
norm.data <- JackStraw(norm.data, num.replicate = 100)
norm.data <- ScoreJackStraw(norm.data, dims = 1:20)

JackStrawPlot(norm.data, dims = 1:15)

ElbowPlot(norm.data)
```

# Cluster the cells using only first four principle components
```{r}
norm.data <- FindNeighbors(norm.data, reduction = "pca", dims = 1:4, k.param=20, compute.SNN=TRUE, force.recalc=TRUE)

norm.data <- FindClusters(norm.data, resolution = 1.3, algorithm=1, random.seed=0)

#Look at cluster IDs of the first 5 cells
head(Idents(norm.data), 5)
clusters_res1.3 <- norm.data$RNA_snn_res.1.3
write.table(clusters_res1.3, "clusters_res1.3_only_Irf1_FF_only_Ikbkb_FF_res_1.3.txt", sep="\t", quote=F)
```


# Running UMAP for clustering 
```{r}

norm.data <- RunUMAP(norm.data, umap.method="uwot", dims = 1:4, seed.use=10,  n.components=2, n.neighbors = 7, spread=1, min.dist=2.5)

DimPlot(norm.data, reduction = "umap", pt.size=2)
```




# Dimensional reduction plot, with cells colored by a quantitative feature
# Feature Plots
# Genes shown in heatmap (Merad paper: fig 1b)
# Not found Sirpa
```{r}
# These genes are not present in our data ("Aldh1a2","Il4i1", "Cd209a","H2-DMb2","Itgam")
gene_list <- c("Ccr7", "Fscn1", "Cd274", "Pdcd1lg2", "Cd200", "Fas", "Il4ra", "Socs2", "Relb", "Xcr1", "Clec9a", "Cadm1", "Naaa","Fscn1","Relb")

for (i in 1:length(gene_list)) {
  print(FeaturePlot(object = norm.data, features = gene_list[i],reduction = "umap",min.cutoff = 0,max.cutoff = 7.5,pt.size=2))
}
```


# Genes for Maturation (Merad paper: fig 1c)
```{r}
gene_list <- c("Cd40", "Il12b", "Ccr7" ,"Ccl5", "Fscn1", "Ccl22", "Cxcl9")


for (i in 1:length(gene_list)) {
  print(FeaturePlot(object = norm.data, features = gene_list[i],reduction = "umap" , min.cutoff = 0, max.cutoff = 7.5, pt.size = 2))
}
```





# Make violin plots
```{r}

gene_list_vln_plt <- c("Irf1","Ikbkb")


for (i in 1:length(gene_list_vln_plt)) {
  print(VlnPlot(object = norm.data, features = gene_list_vln_plt[i],y.max = 10))
}


```



# extract markers for the clusters, and show some of them:
```{r extracting markers for the clusters norm.data2}

 for(i in 0:3) {
  assign(paste("cluster", i,".markers", sep = ""), FindMarkers(norm.data, ident.1 = i, test.use = "bimod", only.pos = TRUE))
    print(paste("cluster", i,".markers", sep = ""))
    print(head(get(paste("cluster", i,".markers", sep = "")),n=20))
    write.table(get(paste("cluster", i,".markers", sep = "")),paste("cluster", i,".markers_only_Irf1_FF_only_Ikbkb_FF_res_1.3.txt", sep = ""),row.names = TRUE)

      }
```




# Make heatmap showing expression of several genes
```{r}




gene_list_all <- c("ANXA3", "AURKA", "AURKB", "BCL2L14", "BHLHE40", "BIRC2", "BIRC5", "CACNB3", "CCL22", "CCL5", "CCNB1", "CCNB2", "CCND1", "CCR7", "CD200", "CD274", "CD40", "CD80", "CD83", "CDK1", "CDK2", "CDK4", "CDKN2D", "CDKN3", "CFLAR", "CKS1B", "CKS2", "CLIC4", "FAM49A", "FAS", "FSCN1", "GADD45B", "GLIPR2", "GNB4", "GPD2", "GYPC", "HTRA2", "ICAM1", "IFI204", "IFIT2", "IFIT3", "IL12B", "IL15", "IL15RA", "IL21R", "IL2RG", "INSL6", "IRF1", "ISG20", "JAK2", "KIF22", "KLF6", "KPNA2", "KTN1", "LAPTM4B", "LBR", "LGMN", "LITAF", "MAP3K8", "MARCKSL1", "MCM5", "MKI67", "MXD1", "MYC", "NFKB2", "NFKBIA", "NFKBIE", "NOSTRIN", "NUDT17", "PCGF5", "PDE4B", "PFKFB3", "PIK3R1", "PLK1", "POGLUT1", "PPP1R15A", "PSME1", "PSME2", "RASA2", "RELB", "RGS1", "RNF115", "SAMHD1", "SAMSN1", "SCIN", "SCPEP1", "SERPINB1A", "SERPINB6B", "SERPINB9", "SMC4", "SOCS2", "SPINT2", "SPPL2A", "SQSTM1", "STAP2", "STAT1", "STAT3", "STAT4", "STMN1", "SWAP70", "TACC3", "TANK", "TMEM123", "TMEM19", "TMEM39A", "TNFAIP3", "TNFSF9", "TOP2A", "TRAF1", "TRAFD1", "TXNDC17",  "VWA5A", "ZMYND15", "USP18", "STAT2","MX1", "ISG15", "IRF7", "IFITM3", "IFITM2", "IFIH1", "BST2", "GBP2", "NCOA7", "CD69",  "Cxcl9", "IL18BP")



gene_list_cell_cycle <- c("STMN1", "CKS1B", "SMC4", "TACC3", "TOP2A", "CDK1", "MCM5", "BIRC5", "AURKB", "LBR", "PLK1", "CKS2", "KIF22", "AURKA", "CDKN3", "MYC", "KPNA2",  "CCNB2", "CDK4", "MKI67", "CCND1",  "CCNB1", "CDKN2D", "CDK2")

gene_list_tnf_via_Nfkb <- c( "ICAM1", "MAP3K8", "CD83","NFKB2", "KLF6",  "TANK", "BIRC2", "RELB", "TNFAIP3", "NFKBIE", "NFKBIA", "IL15RA", "PPP1R15A", "BHLHE40", "SQSTM1", "MXD1", "IRF1", "LITAF", "TNFSF9", "PFKFB3","TRAF1", "CD80", "CFLAR", "GADD45B", "CCL5", "IL12B")


gene_list_inflammatory <- c("ICAM1", "NFKBIA",  "MXD1", "IRF1", "RGS1", "TNFSF9", "CD40", "CCL5", "CCL22", "CCR7", "JAK2",  "SOS1", "STAT3", "STAT1", "IL15",  "IL2RG", "PIK3R1", "IL15RA", "IL21R", "STAT4", "SOCS2", "IL12B")


gene_list_IFNgamma <- c("PSME2", "TRAFD1", "PSME1", "ICAM1",  "SAMHD1", "STAT1", "PDE4B", "SPPL2A","NFKBIA", "IL15RA", "IRF1", "FAS", "CD274", "CD40","CD69", "STAT4", "CCL5","Cxcl9","JAK2","IL18BP","NFKBIA")

gene_list_IFNalpha <- c("GBP2","NCOA7")


gene_list_IFNa_g <- c("BST2","IFIH1","IFIT2","IFITM2","IFITM3","IRF7","ISG15","MX1","STAT2","USP18")



gene_list_maturation <- c("STAT1", "CLIC4",  "TANK", "TNFAIP3", "MARCKSL1", "NFKBIA", "IL15RA", "IRF1", "RGS1", "GPD2", "TNFSF9", "TRAF1", "CD80", "CFLAR", "GADD45B", "SERPINB9", "CD200", "SPINT2", "FAM49A", "GNB4","RASA2",  "GYPC", "BIRC2", "RELB", "POGLUT1", "INSL6", "MARCKSL1", "IL15RA", "LAPTM4B", "HTRA2", "TMEM19", "MXD1",  "SWAP70", "PCGF5", "IRF1", "VWA5A", "SERPINB1A", "LGMN", "FAS", "PFKFB3", "TMEM39A", "BCL2L14", "SCPEP1", "RNF115", "ZMYND15","TXNDC17", "SCIN", "CD274", "CFLAR", "CD40", "STAP2", "IL21R", "KTN1", "NOSTRIN", "TMEM123", "ANXA3", "SERPINB6B", "SERPINB9", "NUDT17", "CCL5", "GLIPR2", "SAMSN1", "CACNB3", "CCR7", "SOCS2", "FSCN1")


 gene_bin_val_df <- data.frame(

as.numeric(gene_list_all %in% gene_list_cell_cycle),
as.numeric(gene_list_all %in% gene_list_tnf_via_Nfkb),
as.numeric(gene_list_all %in% gene_list_inflammatory),
as.numeric(gene_list_all %in% gene_list_IFNgamma),
as.numeric(gene_list_all %in% gene_list_IFNalpha),
as.numeric(gene_list_all %in% gene_list_IFNa_g),
as.numeric(gene_list_all %in% gene_list_maturation)

)


gene_list_all <- tolower(gene_list_all)

firstup <- function(x) {
  substr(x, 1, 1) <- toupper(substr(x, 1, 1))
  x
}




gene_list_all <- firstup(gene_list_all)
exp_dataframe <- data.frame(norm.data@assays$RNA@data)
exp_dataframe_all <- data.frame(t(exp_dataframe[gene_list_all,]))


colnames(gene_bin_val_df) <- c("cell_cycle","tnf_via_Nfkb","inflammatory", "IFNgamma","IFNalpha","IFNa_g","maturation")
rownames(gene_bin_val_df) <- gene_list_all


exp_dataframe_all  <- cbind(exp_dataframe_all,clusters$clusters_res1.3)
colnames(exp_dataframe_all)[128] <- "clusters"

known_levels <- c("C3", "C0", "C2","C1")
my_order <- order(factor(exp_dataframe_all$clusters, levels = known_levels, ordered=TRUE))
exp_dataframe_all  <- exp_dataframe_all[my_order, ]

exp_dataframe_all <- data.frame(t(exp_dataframe_all))
exp_dataframe_all <- exp_dataframe_all[-c(128), ]


exp_mat_all <- as.matrix(exp_dataframe_all)

exp_mat_all <- apply(exp_mat_all, 2, as.numeric)
rownames(exp_mat_all) <- rownames(exp_dataframe_all)



#tiff("test.tiff", units="in", width=20, height=11, res=300)
# insert ggplot code

pheatmap(exp_mat_all,fontsize = 5,
        cluster_rows = T,
         cluster_cols = FALSE,
          legend = T,
          show_rownames = T, 
          show_colnames = T, 
         cellwidth =3, 
         cellheight =5,
         annotation_col= clusters,
      annotation_row = gene_bin_val_df)
```
